### PR TITLE
Arbitrage with router

### PR DESCRIPTION
**Pull request by jepidoptera**
Router doesn't quite implement the whole AMM interface, but it has enough to work for arbitrage.

### CODE REVIEW

**Review by poliwop — CHANGES_REQUESTED** [review]
Mostly looks good, see my comment in test_router.py.. I am not sure we are choosing between withdraw_asset and remove_liquidity correctly.

File: hydradx/model/amm/omnipool_router.py [hunk 1]
```diff
@@ -45,105 +45,288 @@ def price_route(self, tkn: str, denomination: str, tkn_pool_id: str, denom_pool_
...
         return float('inf')
 
+    @property
+    def liquidity(self):
+        return {
+            tkn: sum([exchange.liquidity[tkn] if tkn in exchange.liquidity else 0 for exchange in self.exchanges.values()])
+            for tkn in [tkn for exchange in self.exchanges.values() for tkn in exchange.asset_list]
+        }
+
+    def buy_spot(self, tkn_buy: str, tkn_sell: str, fee: float = -1):
```
[the remark this answers is not in the document]
Comment (jepidoptera) [reply hydradx/model/amm/omnipool_router.py:63]: You're right, that is a better idea. I'll change it.

File: hydradx/model/amm/stableswap_amm.py [hunk 2]
```diff
@@ -473,31 +475,48 @@ def add_liquidity(
         return self
 
     def add_liquidity_spot(self, tkn_add: str):
-        '''Calculates spot price of adding liquidity as shares denominated in liquidity'''
-        trade_size = self.liquidity[tkn_add] / 10**7  # chosen based on Python precision
+        """Calculates spot price of adding liquidity as shares denominated in liquidity"""
+        if type(self.liquidity[tkn_add]) == float or type(self.liquidity[tkn_add]) == int:
```
[the remark this answers is not in the document]
Comment (jepidoptera) [reply hydradx/model/amm/stableswap_amm.py:479]: That is what it's for, and sure

File: hydradx/model/amm/arbitrage_agent_general.py [hunk 3]
```diff
@@ -243,7 +244,7 @@ def calculate_arb_amount(
     buy_ex_max_sell = buy_ex_max_sell if buy_ex_max_sell is not None else float('inf')
     max_buy = buy_ex.calculate_buy_from_sell(
         tkn_buy=buy_ex_tkn_pair[0], tkn_sell=buy_ex_tkn_pair[1], sell_quantity=buy_ex_max_sell
-    )
+    ) if hasattr(buy_ex, 'calculate_buy_from_sell') else float('inf')
```
[the remark this answers is not in the document]
Comment (jepidoptera) [reply hydradx/model/amm/arbitrage_agent_general.py:247]: No. I added logic at line 272 (buy_ex_sold > buy_ex_max_sell) which addresses this without needing a calculate_buy_from_sell function

File: hydradx/model/amm/omnipool_router.py [hunk 4]
```diff
@@ -2,35 +2,35 @@
...
-            if isinstance(exchanges[exchange_id], OmnipoolState):
+        for exchange_id in self.exchanges:
+            if isinstance(self.exchanges[exchange_id], OmnipoolState):
                 if self.omnipool_id is not None:
                     raise ValueError('Multiple Omnipools in exchange list')
                 else:
                     self.omnipool_id = exchange_id
         if self.omnipool_id is None:
             raise ValueError('No Omnipool in exchange list')
+        self.omnipool: OmnipoolState = self.exchanges[self.omnipool_id]
```
[the remark this answers is not in the document]
Comment (jepidoptera) [reply hydradx/model/amm/omnipool_router.py:23]: Yeah, maybe so. All this does is facilitate intellisense completions and I guess save a few dictionary lookups

**Review by poliwop — APPROVED** [review]
Looks good.

**Pull request merged**